homeassistant/helpers/restore_state.py

```diff
@@ -178,30 +178,29 @@ def async_get_stored_states(self) -> list[StoredState]:
         now = dt_util.utcnow()
         all_states = self.hass.states.async_all()
         # Entities currently backed by an entity object
-        current_entity_ids = {
-            state.entity_id
+        current_states_by_entity_id = {
+            state.entity_id: state
             for state in all_states
             if not state.attributes.get(ATTR_RESTORED)
         }
 
         # Start with the currently registered states
         stored_states = [
             StoredState(
-                state, self.entities[state.entity_id].extra_restore_state_data, now
+                current_states_by_entity_id[entity_id],
+                entity.extra_restore_state_data,
+                now,
             )
-            for state in all_states
-            if state.entity_id in self.entities
-            and
-            # Ignore all states that are entity registry placeholders
-            not state.attributes.get(ATTR_RESTORED)
```

Comment:
Checking `not state.attributes.get(ATTR_RESTORED)` is now avoided since this list is always a subset of `current_states_by_entity_id`